### PR TITLE
Fix l1 fat event filters 90x bis - now use TCDS unpacked data for L1T

### PR DESCRIPTION
**Pull request by rekovic**
This is a 90x version of 81x PR #16782 (and also replacement for PR #16799)

The two commits 	70ae7c3 and 	b5b3430 are to address @davidlange6 comment https://github.com/cms-sw/cmssw/pull/16799#discussion_r90223960
Will cherry-pick them  to 81x PR once this merged.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc [hunk 1]
```diff
@@ -68,12 +65,11 @@ class L1TValidationEventFilter : public edm::EDFilter {
 // constructors and destructor
 //
 L1TValidationEventFilter::L1TValidationEventFilter(const edm::ParameterSet& iConfig) :
+  src_(iConfig.getParameter<edm::InputTag>("src")),
   period_( iConfig.getUntrackedParameter<int>("period", 107) )
 {
   //now do what ever initialization is needed
 
-  fedData_ = consumes<FEDRawDataCollection>(iConfig.getParameter<edm::InputTag>("inputTag"));
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: why did you remove the `consumes` call ?

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc [hunk 2]
```diff
@@ -68,12 +65,11 @@ class L1TValidationEventFilter : public edm::EDFilter {
 // constructors and destructor
 //
 L1TValidationEventFilter::L1TValidationEventFilter(const edm::ParameterSet& iConfig) :
+  src_(iConfig.getParameter<edm::InputTag>("src")),
   period_( iConfig.getUntrackedParameter<int>("period", 107) )
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: unrelated to this PR, but this should be a `tracked` parameter

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc [hunk 3]
```diff
@@ -95,22 +91,15 @@ bool
...
 {
   using namespace edm;
-  
-  edm::Handle<FEDRawDataCollection> feds;
-  iEvent.getByToken(fedData_, feds);
 
-  if (!feds.isValid()) {
-    LogError("L1T") << "Cannot unpack: no FEDRawDataCollection found";
+  Handle<int> triggerCount;
+  iEvent.getByLabel(InputTag(src_.label(),"triggerCount"), triggerCount);
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: It's usually preferable to let the configuration specify the whole InputTag, not only the module name with a hardcoded instance label.

**Review by fwyzard — CHANGES_REQUESTED** [review]

**Comment by fwyzard** [comment]
+1

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc (hunk 1 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: Why keep `consumes` if `fedData_` is not used anymore?

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc (hunk 3 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: In principle it can all be configurable, but the intention was to only have the label configurable to match the _unpacker_  module label used. In this case the product <int> of instance label "triggerCount" is of interest.  This instance label is assigned to the product by the TcdsRawToDigi _unpacker_ and it differentiates it from other <int> product put in Event.

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc (hunk 1 above)
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: sorry for not being very clear - I meant that
```
edm::InputTag src_;
...
src_(iConfig.getParameter<edm::InputTag>("src")),
```
should use the `consumes` interface, and rather be
```
edm::EDGetTokenT<int> src_;
...
src_(consumes<int>(iConfig.getParameter<edm::InputTag>("src"));
```

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc (hunk 3 above)
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: The reason for having it configurable is that if you write a different producer for it, you may have a different label.

Since the most common use case is to set the input tag in a cfi fie (or even in the fillDescriptions method) and always use the default, users won;t have to type the extra label anyway.

File: EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc (hunk 1 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/L1TValidationEventFilter.cc]: In principle, you are right, `getByToken` should be faster than `getByLabel`.  Can switch to it.

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
please test

**Comment by fwyzard** [comment]
+1

File: EventFilter/Utilities/plugins/TcdsRawToDigi.cc [hunk 4]
```diff
@@ -87,20 +88,25 @@ void TcdsRawToDigi::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
     iEvent.getByToken(dataToken_,rawdata);
 
     int nibble=-99;
+    int triggerCount = -99;
     if( rawdata.isValid() ) {
         const FEDRawData& tcdsData = rawdata->FEDData(FEDNumbering::MINTCDSuTCAFEDID);
         if(tcdsData.size()>0){
             evf::evtn::TCDSRecord tcdsRecord(tcdsData.data());
             nibble = (int)tcdsRecord.getHeader().getData().header.nibble;
+            triggerCount = (int)tcdsRecord.getHeader().getData().header.triggerCount;
```
Comment (slava77) [inline EventFilter/Utilities/plugins/TcdsRawToDigi.cc]: triggerCount is apparently uint64_t
https://twiki.cern.ch/twiki/bin/view/CMS/TcdsEventRecord
https://github.com/cms-sw/cmssw/blob/CMSSW_8_2_0/EventFilter/FEDInterface/interface/FED1024.h#L72

it seems inappropriate to save it as an int

Comment (slava77) [inline EventFilter/Utilities/plugins/TcdsRawToDigi.cc]: if negative numbers should remain useful, I suppose int64_t should still work, since the normal trigger count will not go over 50 bits.

It's still worth to make a proper masking of the high bit (sign).

Comment (rekovic) [inline EventFilter/Utilities/plugins/TcdsRawToDigi.cc]: @slava77  Thanks!  Will use int64_t, as to keep the sign in the unpacker.

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17082 744b16b
- reco (somewhat) related part of the change is to unpack triggerCount from the TCDS header
- jenkins tests pass and comparisons show no differences
- inspection of a copy of the jenkins output from 136.731 confirms presence of the new product ``` long_tcdsDigis_triggerCount_reRECO ```

```
e->Scan("EventAuxiliary.run():EventAuxiliary.luminosityBlock():EventAuxiliary.event():\
long_tcdsDigis_triggerCount_reRECO.obj:int_tcdsDigis_nibble_reRECO.obj")
************************************************************************
*    Row   * EventAuxi * EventAuxi * EventAuxi * long_tcds * int_tcdsD *
************************************************************************
*        0 *    274199 *        21 *  39389360 *   1858229 *        27 *
*        1 *    274199 *        21 *  38626099 *   1094968 *         3 *
*        2 *    274199 *        21 *  39797590 *   2266459 *        40 *
*        3 *    274199 *        21 *  40112763 *   2581632 *        50 *
*        4 *    274199 *        21 *  39146340 *   1615209 *        20 *
*        5 *    274199 *        21 *  39564069 *   2032938 *        33 *
*        6 *    274199 *        21 *  39150144 *   1619013 *        20 *
*        7 *    274199 *        21 *  40038720 *   2507589 *        48 *
*        8 *    274199 *        21 *  40351994 *   2820863 *        58 *
*        9 *    274199 *        21 *  38618538 *   1087407 *         3 *
```

**Comment by fwyzard** [comment]
+1

**Comment by rekovic** [comment]
@jimbrooke this PR is of interest for you.
